File: services/gateway/app/middleware/rbac.ts
/**
 * Server-side RBAC enforcement (gateway).
 * ----------------------------------------------------------------------------
 * Mirrors the frontend's authStore.has(perm) check, but on the SERVER and
 * against MySQL — so the server never trusts the client's claimed
 * permissions.
 *
 * Design:
 *   - Identity is taken from req.user (set by the existing authMiddleware
 *     from the JWT). We use only the `uid` claim as a key into MySQL.
 *   - Permissions are looked up live from `users → user_roles →
 *     role_permissions → permissions` with a 30-second in-process LRU.
 *   - `requirePermission(perm)` and `requireAnyPermission(...perms)` return
 *     Express middlewares that 403 on missing perms, 401 on missing auth.
 *   - On MySQL outage we fail CLOSED for mutating routes (deny) and OPEN
 *     for read-only routes (allow with a warning). Mutating routes are
 *     identified by HTTP method.
 *
 * The middleware that auto-applies the registry is also exported as
 * `enforceRegistry` and is wired up in main.ts AFTER authMiddleware.
 */

import { Request, Response, NextFunction } from 'express';
import { match as pathToRegexpMatch } from 'path-to-regexp';
import { getPool } from '../db/mysqlPool';
import { logger } from '../utils/logger';
import { JwtPayload } from './auth';
import {
  PERMISSION_REQUIREMENTS_WRITE,
  PERMISSION_REQUIREMENTS_READ,
  RouteRule,
} from './permissions';

// ─── Tiny LRU cache (no external dependency) ─────────────────────────────────

interface CacheEntry {
  perms: Set<string>;
  expires: number;
}

const PERMISSION_CACHE_TTL_MS = 30_000; // 30s — short enough that role
                                        // changes propagate quickly
const PERMISSION_CACHE_MAX = 500;       // cap per-process memory
const cache = new Map<number, CacheEntry>();

function cacheGet(uid: number): Set<string> | undefined {
  const hit = cache.get(uid);
  if (!hit) return undefined;
  if (hit.expires < Date.now()) {
    cache.delete(uid);
    return undefined;
  }
  // touch for LRU semantics
  cache.delete(uid);
  cache.set(uid, hit);
  return hit.perms;
}

function cacheSet(uid: number, perms: Set<string>): void {
  if (cache.size >= PERMISSION_CACHE_MAX) {
    const first = cache.keys().next().value;
    if (first !== undefined) cache.delete(first);
  }
  cache.set(uid, { perms, expires: Date.now() + PERMISSION_CACHE_TTL_MS });
}

/** Test hook: reset the cache between tests. */
export function _clearRbacCache(): void {
  cache.clear();
}

// ─── DB lookup ───────────────────────────────────────────────────────────────

/**
 * Load the permission set for `uid` from MySQL, with LRU caching. Throws on
 * DB error; the caller decides whether to fail open or closed.
 */
export async function loadPermissionsForUser(uid: number): Promise<Set<string>> {
  const cached = cacheGet(uid);
  if (cached) return cached;

  const [rows] = await getPool().query(
    `SELECT DISTINCT p.name
       FROM permissions p
       JOIN role_permissions rp ON rp.permission_id = p.id
       JOIN user_roles ur        ON ur.role_id = rp.role_id
      WHERE ur.user_id = ?`,
    [uid],
  );
  // mysql2 typing: rows is RowDataPacket[]; cast for our narrow shape.
  const perms = new Set<string>(
    (rows as unknown as Array<{ name: string }>).map((r) => r.name),
  );
  cacheSet(uid, perms);
  return perms;
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

interface ReqWithUser extends Request {
  user?: JwtPayload;
  id?: string;
}

function getIdentity(req: ReqWithUser): { uid: number | null; sub: string | null } {
  const u = req.user;
  if (!u) return { uid: null, sub: null };
  // API-key requests synthesize a user with sub like "apikey:..." but no uid;
  // we treat them as super-user for RBAC because they already authenticated
  // out-of-band. (Same posture as authMiddleware.)
  if (typeof u.sub === 'string' && u.sub.startsWith('apikey:')) {
    return { uid: -1, sub: u.sub }; // sentinel: -1 means "skip RBAC"
  }
  return {
    uid: typeof u.uid === 'number' ? u.uid : null,
    sub: typeof u.sub === 'string' ? u.sub : null,
  };
}

function deny(
  res: Response,
  status: 401 | 403,
  message: string,
  traceId: string | undefined,
  missing?: string[],
): void {
  res.status(status).json({
    error: message,
    code: status === 401 ? 'AUTH_FAILED' : 'FORBIDDEN',
    missing,
    trace_id: traceId,
  });
}

const MUTATING_METHODS = new Set(['POST', 'PUT', 'PATCH', 'DELETE']);

// ─── Public API: per-route factories ────────────────────────────────────────

/**
 * Require the user to hold ALL of the named permissions. Use this for
 * obvious single-permission gates: requirePermission('agents.write').
 */
export function requirePermission(...needed: string[]) {
  return async function rbacRequirePermission(
    req: ReqWithUser,
    res: Response,
    next: NextFunction,
  ): Promise<void> {
    const traceId = req.id;
    const { uid, sub } = getIdentity(req);

    if (uid === null && sub === null) {
      deny(res, 401, 'Not authenticated', traceId);
      return;
    }
    if (uid === -1) {
      // API-key — already authenticated, bypass RBAC
      next();
      return;
    }
    if (uid === null) {
      // JWT without uid (legacy) — treat as auth failure
      deny(res, 401, 'Token missing user identifier', traceId);
      return;
    }

    let perms: Set<string>;
    try {
      perms = await loadPermissionsForUser(uid);
    } catch (err) {
      const reason = err instanceof Error ? err.message : String(err);
      logger.error('rbac_db_unavailable', {
        layer: 'middleware',
        trace_id: traceId,
        sub,
        reason,
        method: req.method,
        path: req.path,
      });
      // Mutating route → fail closed; read route → fail open
      if (MUTATING_METHODS.has(req.method.toUpperCase())) {
        deny(res, 403, 'RBAC unavailable; denying mutation', traceId);
        return;
      }
      logger.warn('rbac_fail_open_read', {
        layer: 'middleware',
        trace_id: traceId,
        sub,
        method: req.method,
        path: req.path,
      });
      next();
      return;
    }

    const missing = needed.filter((p) => !perms.has(p));
    if (missing.length > 0) {
      logger.warn('rbac_denied', {
        layer: 'middleware',
        trace_id: traceId,
        sub,
        missing,
        method: req.method,
        path: req.path,
      });
      deny(res, 403, `Access denied — missing permission: ${missing.join(', ')}`, traceId, missing);
      return;
    }
    next();
  };
}

/**
 * Require AT LEAST ONE of the named permissions. Useful for read-paths
 * shared across multiple roles (e.g. governance.read OR ml_insights.read).
 */
export function requireAnyPermission(...needed: string[]) {
  return async function rbacRequireAnyPermission(
    req: ReqWithUser,
    res: Response,
    next: NextFunction,
  ): Promise<void> {
    const traceId = req.id;
    const { uid, sub } = getIdentity(req);

    if (uid === null && sub === null) {
      deny(res, 401, 'Not authenticated', traceId);
      return;
    }
    if (uid === -1) {
      next();
      return;
    }
    if (uid === null) {
      deny(res, 401, 'Token missing user identifier', traceId);
      return;
    }

    let perms: Set<string>;
    try {
      perms = await loadPermissionsForUser(uid);
    } catch (err) {
      const reason = err instanceof Error ? err.message : String(err);
      logger.error('rbac_db_unavailable', {
        layer: 'middleware',
        trace_id: traceId,
        sub,
        reason,
        method: req.method,
        path: req.path,
      });
      if (MUTATING_METHODS.has(req.method.toUpperCase())) {
        deny(res, 403, 'RBAC unavailable; denying mutation', traceId);
        return;
      }
      logger.warn('rbac_fail_open_read', {
        layer: 'middleware',
        trace_id: traceId,
        sub,
        method: req.method,
        path: req.path,
      });
      next();
      return;
    }

    if (!needed.some((p) => perms.has(p))) {
      logger.warn('rbac_denied_any', {
        layer: 'middleware',
        trace_id: traceId,
        sub,
        anyOf: needed,
        method: req.method,
        path: req.path,
      });
      deny(res, 403, `Access denied — requires one of: ${needed.join(', ')}`, traceId, needed);
      return;
    }
    next();
  };
}

// ─── Registry-based auto-enforcement ─────────────────────────────────────────

interface CompiledRule extends RouteRule {
  matcher: ReturnType<typeof pathToRegexpMatch>;
}

function compile(rules: ReadonlyArray<RouteRule>): CompiledRule[] {
  return rules.map((r) => ({
    ...r,
    matcher: pathToRegexpMatch(r.path),
  }));
}

const COMPILED_WRITE = compile(PERMISSION_REQUIREMENTS_WRITE);
const COMPILED_READ = compile(PERMISSION_REQUIREMENTS_READ);

function findRule(method: string, path: string): RouteRule | undefined {
  const upper = method.toUpperCase() as RouteRule['methods'][number];
  for (const r of [...COMPILED_WRITE, ...COMPILED_READ]) {
    if (r.methods.length === 0 || r.methods.includes(upper)) {
      if (r.matcher(path)) return r;
    }
  }
  return undefined;
}

/**
 * Express middleware: looks at req.method + req.path against the central
 * registry and enforces the matching rule. Mount this AFTER authMiddleware
 * but BEFORE the per-resource routers.
 *
 * If no rule matches, the request is allowed through — explicit gating only.
 */
export function enforceRegistry(
  req: ReqWithUser,
  res: Response,
  next: NextFunction,
): void {
  // Skip enforcement entirely for the auth flow itself
  if (req.path.startsWith('/auth/')) {
    next();
    return;
  }
  const rule = findRule(req.method, req.path);
  if (!rule) {
    next();
    return;
  }
  // Delegate to the per-permission middleware so logging is consistent
  void requireAnyPermission(...rule.anyOf)(req, res, next);
}

File: services/orchestrator/app/middleware/rbac.py
"""Server-side RBAC enforcement (orchestrator).

Companion to ``services/gateway/app/middleware/rbac.ts``. Provides FastAPI
``Depends`` factories that look up the caller's permission set from MySQL
and 403 if the required permission is missing.

Identity source
---------------
The gateway is the trust boundary: it verifies the JWT and forwards the
resolved user identity to downstream services as the ``x-user-id`` header
(numeric DB id). The orchestrator does NOT re-verify the JWT — it just
trusts the gateway-set header. If a request arrives with no ``x-user-id``,
we treat it as unauthenticated (401) on a gated route.

Failure mode
------------
- MySQL unavailable + mutating route (POST/PUT/PATCH/DELETE)
    → fail CLOSED (deny).
- MySQL unavailable + read-only route (GET/HEAD)
    → fail OPEN with a warning so the system stays observable.
- Missing ``x-user-id`` on a gated route → 401 (not 403).
- ``x-api-key`` set with no ``x-user-id`` → bypass RBAC. The gateway's
  authMiddleware already vetted the API key.

Caching
-------
A 30-second in-process LRU keeps DB round-trips off the hot path. Bypass
the cache via ``_clear_rbac_cache()`` from tests.
"""

from __future__ import annotations

import threading
import time
from typing import Callable, Dict, Optional, Set, Tuple

import mysql.connector
from fastapi import HTTPException, Request

from app.config import settings
from app.utils.logger import logger

# ---------------------------------------------------------------------------
# Cache
# ---------------------------------------------------------------------------
_PERMISSION_CACHE_TTL_SEC = 30.0
_PERMISSION_CACHE_MAX = 500
_cache_lock = threading.Lock()
# uid → (perms, expires_at)
_cache: Dict[int, Tuple[Set[str], float]] = {}


def _cache_get(uid: int) -> Optional[Set[str]]:
    with _cache_lock:
        hit = _cache.get(uid)
        if not hit:
            return None
        perms, expires = hit
        if expires < time.monotonic():
            _cache.pop(uid, None)
            return None
        # Touch for LRU
        _cache.pop(uid, None)
        _cache[uid] = (perms, expires)
        return perms


def _cache_set(uid: int, perms: Set[str]) -> None:
    with _cache_lock:
        if len(_cache) >= _PERMISSION_CACHE_MAX:
            # Evict oldest
            oldest_key = next(iter(_cache))
            _cache.pop(oldest_key, None)
        _cache[uid] = (perms, time.monotonic() + _PERMISSION_CACHE_TTL_SEC)


def _clear_rbac_cache() -> None:
    """Test hook."""
    with _cache_lock:
        _cache.clear()


# ---------------------------------------------------------------------------
# DB lookup
# ---------------------------------------------------------------------------
def _conn():
    """Same shape as routes/catalog.py:_conn() so we hit one place to swap."""
    return mysql.connector.connect(
        host=settings.mysql_host,
        port=settings.mysql_port,
        user=settings.mysql_user,
        password=settings.mysql_password,
        database=settings.mysql_db,
        connection_timeout=5,
    )


def load_permissions_for_user(uid: int) -> Set[str]:
    """Read the user's permission set from MySQL, with caching.

    Raises on DB error so the caller can decide on the fail-open / fail-closed
    posture per route.
    """
    cached = _cache_get(uid)
    if cached is not None:
        return cached

    conn = _conn()
    try:
        cur = conn.cursor()
        cur.execute(
            """
            SELECT DISTINCT p.name
              FROM permissions p
              JOIN role_permissions rp ON rp.permission_id = p.id
              JOIN user_roles ur        ON ur.role_id = rp.role_id
             WHERE ur.user_id = %s
            """,
            (uid,),
        )
        rows = cur.fetchall() or []
    finally:
        try:
            conn.close()
        except Exception:
            pass
    perms = {r[0] for r in rows if r and r[0]}
    _cache_set(uid, perms)
    return perms


# ---------------------------------------------------------------------------
# Identity extraction
# ---------------------------------------------------------------------------
def _extract_uid(request: Request) -> Optional[int]:
    """Return the caller's numeric user id, or ``None`` if absent.

    A returned value of ``-1`` is the "API-key bypass" sentinel — same
    convention as the gateway middleware.
    """
    if request.headers.get("x-api-key"):
        return -1
    raw = request.headers.get("x-user-id")
    if not raw:
        return None
    try:
        return int(raw)
    except (TypeError, ValueError):
        return None


# ---------------------------------------------------------------------------
# Public API: dependency factories
# ---------------------------------------------------------------------------
_MUTATING = {"POST", "PUT", "PATCH", "DELETE"}


def _trace_id(request: Request) -> str:
    return request.headers.get("x-request-id") or "no-trace"


def require_permission(perm: str) -> Callable[[Request], None]:
    """FastAPI ``Depends`` factory: requires the caller to hold ``perm``.

    Usage::

        @router.post(
            "/crawlers",
            dependencies=[Depends(require_permission("catalog.write"))],
        )
        async def create_crawler(...): ...
    """

    def _dep(request: Request) -> None:
        _enforce(request, any_of=(perm,))

    _dep.__name__ = f"require_permission_{perm.replace('.', '_')}"
    return _dep


def require_any_permission(*perms: str) -> Callable[[Request], None]:
    """FastAPI ``Depends`` factory: caller must hold at least one of perms."""

    def _dep(request: Request) -> None:
        _enforce(request, any_of=perms)

    _dep.__name__ = "require_any_permission_" + "_".join(p.replace(".", "_") for p in perms)
    return _dep


# ---------------------------------------------------------------------------
# Core enforcement
# ---------------------------------------------------------------------------
def _enforce(request: Request, any_of: Tuple[str, ...]) -> None:
    trace_id = _trace_id(request)
    method = (request.method or "GET").upper()
    is_mutating = method in _MUTATING

    uid = _extract_uid(request)

    if uid is None:
        # No identity at all — 401 even on read routes.
        logger.warning(
            "rbac_no_identity",
            layer="middleware",
            method=method,
            path=str(request.url.path),
            trace_id=trace_id,
        )
        raise HTTPException(
            status_code=401,
            detail={
                "error": "Not authenticated; missing x-user-id",
                "code": "AUTH_FAILED",
                "trace_id": trace_id,
            },
        )

    if uid == -1:
        # API-key bypass — already authenticated by the gateway.
        return

    try:
        perms = load_permissions_for_user(uid)
    except mysql.connector.Error as exc:
        logger.error(
            "rbac_db_unavailable",
            layer="middleware",
            method=method,
            path=str(request.url.path),
            trace_id=trace_id,
            uid=uid,
            error=str(exc),
        )
        if is_mutating:
            raise HTTPException(
                status_code=403,
                detail={
                    "error": "RBAC unavailable; denying mutation",
                    "code": "FORBIDDEN",
                    "trace_id": trace_id,
                },
            )
        # Read route → fail open
        logger.warning(
            "rbac_fail_open_read",
            layer="middleware",
            method=method,
            path=str(request.url.path),
            trace_id=trace_id,
            uid=uid,
        )
        return

    if not any(p in perms for p in any_of):
        logger.warning(
            "rbac_denied",
            layer="middleware",
            method=method,
            path=str(request.url.path),
            trace_id=trace_id,
            uid=uid,
            any_of=list(any_of),
        )
        raise HTTPException(
            status_code=403,
            detail={
                "error": f"Access denied — requires one of: {', '.join(any_of)}",
                "code": "FORBIDDEN",
                "missing": list(any_of),
                "trace_id": trace_id,
            },
        )
